### ISSUE TITLE
Choose a pipeline for mlflow runs

### ISSUE THREAD

**Issue by NoMoreFriends** [Galileo-Galilei/kedro-mlflow#92]
## Description
I have been working with kedro for a long time now, tried to make it a standard for ML projects in our team, I had to use mlflow inside the nodes first, then the hooks came out which was a great feature. Now that we have kedro-mlflow, I juste read about it a bit, and in your example you did not specify which pipeline should trigger the mlflow start experiment. 

## Context
I think it is very important, because as you know, in a real project there is a lot of pipelines from data preparation to statistical modeling. We should only use mlflow for the training part.

## Possible Implementation
Unless I missed something in the documentation, the easiest way to do it in my opinion is to inherit from the `MlflowNodeHook`, `MlflowPipelineHook` classes to specify which nodes and pipelines should trigger the mlflow experiment launch. Is there an easier way through the yaml files or the pipelines definition ?

PS: **Thank you for this amazing framework** 👍

**Comment by Galileo-Galilei** [comment]
Hi @NoMoreFriends, thanks for your feedback!

To be honest, this is a **very common feature request**. There are several reasons I did not implemented it yet, some "philosophical" and some "technical".

### "Philosophical" reasons

As you mention, machine learning projects always have several pipelines, and some are hardly related to the "training" part.

A common workflow is to have:
1. An "etl" applications which creates "instances" (e.g. a `pandas.DataFrame` which contains several examples of the "business object" your ml applications will handle (for example a picture, an iris flower description, a movie review...) and "labels".
2. A "ml" application which takes the instances as input and return predictions
3. An "user oriented" applications which takes instances and your ML model and "consumes" it to make some actions


In your Kedro project, these different applications are themselves divided into several pipelines:
- the "etl" might be have a pipeline to create the training dataset (becase you extract a given period in the past) and another to create the "live" data for prediction
- the "ml" has at least a "training" and an "inference" pipeline, and maybe some other pipelines to explain the predictions (with shap values, activation maps...), one to monitor predictions...
- the "user" app is sometimes partly develop by the datascientist (a shiny / dash / Tableau/ streamlit dashboard to demonstrate the value for instance), and will likely be outsourced to a third party (the responsibilty of using the model will likely belong to the "real" business app)

While designing the model, the 3 apps falls under the development of the data scientists because he/she needs to iterate fast. In production, these apps will be orchestrated / scheduled independently and sometimes (often?) a different team will be responsible for each app (a team for the etl and a team for the user app?).

It turned out that many users who asked for "disabling" mlflow automatic versioning in their datascience projects finally acknowledged that it is useful to at least version parameters for all the apps and not only the ml ones. Typical example when this is needed for reproducibility includes:
- the etl app creates the training dataset with a SQL request over a time period. **During experimentation, the model is trained with different time period** (to check stability for instance): **if the period parameter is not recorded, it is very hard to remember on which training dataset a given model was trained.**
- the user app shows the results of several models (for A/B testing) and you want to version the predictions to comapre the results depending on users actions.

These two use cases are no longer prevalent in production because you will likely not change the trainnig dataset definition often, and versioning predictions will no longer be the responsibility of the data scientist. Moreover, you have better tools to handle such uses cases (e.g. DVC for the first, Elastic/Kibana for the second) but mlflow can handle it as well.

As a consequence, **I often figured out that in experimentation phase it is often useful to version all runs** (and not costly at all if you version only parameters and not data, while lack of versioning may be hurtful), while in production these will be different projects (I mean, different Kedro projects, with different ProjectContext, and likely orchestrated by another tool like airflow), and you can just ignore kedro-mlflow for the projects which do no need it.

### Technical reasons

- Until today and the PR #91, `kedro-mlflow` was not able to make the 2 mlflow hooks interoperate easily. You need to access the name of the current run anywhere you make an mlflow call (like logging), because if you don't, mlflow will automatically start a new run with a default local configuration (and create a mlruns folder in your working directory). Not starting the run in MlflowPipelineHook is not enough to avoid mlflow logging.
- If a user make a call to mlflow inside a node (and the plugin is totally unaware of it - this should never happen in the final version of the plugin, but a few use case are notcovered by the plugin yet), mlflow will use its default configuration and completely mess things up. Forcing the configuration is a "safety guard".

### Design decision and plugin roadmap

Despite above reasons, I totally acknowledge that you sometimes need / want to turn off mlflow logging and you may have some pipelines that do not require mlflow at all. Above all, many users require this feature request and @takikadiri insist that it is very useful for transition to production. As a consequence, **we plan to add this feature in a near future (i.e. within few months)**. However, I have not defined yet what is the best way to do so. 

First, I think that versioning must be activated by default (after all, this is why people install kedro-mlflow). We wil provide an *opt-ou* strategy rather than an *opt-in* strategy. This means we will provide a list of pipelines to *exclude* rather than a list of pipelines to *include* in mlflow versioning. 

Second, we should define how to parameterize run versioning deactivation. The **`mlflow.yml` is a good candidate, but it is becoming overcrowded. 

```yaml
disabled_pipelines: # find a  better name
    - pipeline1
    - __default__
```

Another (or complementary) possibility would be to use extra run params to enable deactivate at run time. Something like:

```bash
kedro run --pipeline=my_pipeline --no-mlflow
```

but it will be hacky while https://github.com/quantumblacklabs/kedro/issues/382 is not solved.

### Short term solution

Unfortunately, I don't have an easy solution to opt out only some pipelines. You can inherit from the two hooks as you suggest, but you will have to manage carefully configuration retrieval. If you're in a hurry, feel free to open a PR. 

You will need to:
- add a key in the mlflow.yml
- update framework/context/config KedroMlflowConfig to manage this new key
- update the hooks:

Before this line : 

https://github.com/Galileo-Galilei/kedro-mlflow/blob/e8f9f09d01cca9a6b9f99006ad482988319bb297/kedro_mlflow/framework/hooks/node_hook.py#L77

Add:

```python
config = get_mlflow_config(self.context)
# not sure where the current running pipeline name is stored
if context.pipeline_name not in config.disabled_pipelines:
   <rest of the code>
```

And do the same in the `MlflowPipelineHook`.

**Comment by Galileo-Galilei** [comment]
Hi @NoMoreFriends, 

sorry for the long radio silence. The functionality described in this issue is partially implemented in the PR #184. 

You can add the following key in your `mlflow.yml`:

```yaml
disable_tracking:
  pipelines: 
    - pipeline_to_disable
```

Some disclaimers:
- it will turn off automatic parameters logging as well as catalog logging (`Mlflow<Artifact|Metrics|Model>DataSets`) from `kedro-mlflow` , but if you have added mlflow code inside your own code (e.g. add `mlflow.log_metric` in one of your node) mlflow will open a new run by default in a mlruns folder at the root of your project (mlflow does not support tracking deactivation).
- This is only compatible with `kedro>=0.17.1`.
- It is not possible yet to add the `--no-mlflow` flag in the cli, but it may become availble in future kedro version since the kedro team has introduced the abilities to override the `run` command in their develop branch but it is not released yet.
- It lacks some tests and documentation, so I will not released the feature before weeks, but feel free to try it out and give feedback!
